apps/examples/webserver_test/webserver_test_main.c

```diff
@@ -0,0 +1,999 @@
+
+	if (param->tls && ws_test_tls_init(client_tls, &param->ssl_config)) {
+		PRNT("Fail to client tls init");
+		goto errout_before_tlsinit;
+	}
+
+	/* Re-initialize portions of the state structure that could have
+	 * been left from the previous time through the loop and should not
+	 * persist with the new connection.
+	 */
+	ws.httpstatus = HTTPSTATUS_NONE;
+	ws.offset = 0;
+
+	if ((sndlen = ws_test_msg_construct(ws.buffer, param, &ws)) < 0) {
+		PRNT("ERROR: construction message failed");
+		goto errout_before_tlsinit;
+	}
+
+	PRNT("sndlen:%d , param->buffer is:\n%s", sndlen,param->buffer);
+
+retry:
+	if ((sockfd = ws_test_socket_connect(&ws)) < 0) {
+		PRNT("ERROR: socket failed: %d", errno);
+		goto errout_before_tlsinit;
+	}
+
+	client_tls->client_fd = sockfd;
+	if (param->tls && (ret = ws_test_tls_handshake(client_tls, ws.hostname))) {
+		if (handshake_retry-- > 0) {
+			if (ret == MBEDTLS_ERR_NET_SEND_FAILED ||
+					ret == MBEDTLS_ERR_NET_RECV_FAILED ||
+					ret == MBEDTLS_ERR_SSL_CONN_EOF) {
+				PRNT("Handshake again.... ");
+				mbedtls_net_free(&(client_tls->tls_client_fd));
+				mbedtls_ssl_free(&(client_tls->tls_ssl));
+				close(sockfd);
+				goto retry;
+			}
+		}
+		goto errout;
+	}
+
+	send_len = sndlen;
+	for (int i = 0; i <= param->keep_alive_count; i++) {
+
+		if (!param->encoding) {
+			buf_len = 0;
+			sndlen = send_len;
+			while (sndlen > 0) {
+				if (param->tls) {
+					ret = mbedtls_ssl_write(&(client_tls->tls_ssl),
+											(const unsigned char *)param->buffer + buf_len,
+											sndlen);
+				} else {
+					ret = send(sockfd, param->buffer + buf_len, sndlen, 0);
+				}
+				if (ret < 1) {
+					PRNT("ERROR: send failed: %d", ret);
+					goto errout;
+				} else {
+					sndlen -= ret;
+					buf_len += ret;
+					PRNT("SEND SUCCESS: send %d bytes", ret);
+				}
+			}
+		} else {
+			ws_test_chunk_transfer(param, send_len, sockfd, client_tls);
+		}
+
+		if (param->callback && param->response == NULL) {
+			param->response = &response;
+			if (ws_test_http_client_response_init(param->response) < 0) {
+				PRNT("ERROR: response init failed: %d", ret);
+				param->response = NULL;
+				goto errout;
+			}
+		}
+		int loopcount = 0;
+		while (!read_finish) {
+			PRNT("Receive start");
+			memset(param->response->message, 0, WS_TEST_CONF_MAX_MESSAGE_SIZE);
+			if (param->tls) {
+				len = mbedtls_ssl_read(&(client_tls->tls_ssl),
+									   (unsigned char *)param->response->message,
+									   WS_TEST_CONF_MAX_MESSAGE_SIZE);
+			} else {
+				len = recv(sockfd, param->response->message,
+						   WS_TEST_CONF_MAX_MESSAGE_SIZE, 0);
+			}
+
+			if (len < 0) {
+				PRNT("Error: Receive Fail %d", len);
+				goto errout;
+			} else if (len == 0) {
+				PRNT("Finish read");
+				if (mlen.message_len - mlen.sentence_start == mlen.content_len) {
+					PRNT("download completed successfully");
+					break;
+				} else {
+					PRNT("Error: Receive Fail");
+					goto errout;
+				}
+			}
+
+			usleep(1);
```

Comment:
Why did you add sleep in here. Adding sleep can cause potential issues. So please check why usleep is needed here. or add comment to code.